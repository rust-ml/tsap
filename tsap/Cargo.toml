[package]
name = "tsap"
version = "0.1.0"
edition = "2018"

[[bin]]
name = "example"
path = "src/example.rs"

[dependencies]
thiserror = "1.0"
tsap_macro = { version = "0.1.0", path = "../tsap_macro" }
toml = { version = "0.5", optional = true }
serde = { version = "1.0", optional = true, features = ["derive"] }

[features]
default = ["toml"]
toml = ["dep:serde", "dep:toml"]
