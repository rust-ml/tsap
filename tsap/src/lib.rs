extern crate tsap_macro;

pub use tsap_macro::param;

#[cfg(feature = "toml")]
pub mod toml_builder;
#[cfg(feature = "toml")]
pub mod templates;

mod error;

pub use error::{Result, Error};

pub trait ParamGuard {
    type Error;

    fn check(&self) -> std::result::Result<(), Self::Error> {
        Ok(())
    }
}

/// Caller semantic to accept owned values and closures in builder pattern
pub struct Call<T>(Box<dyn Fn(T) -> T>);

impl<T> Call<T> {
    pub fn call(self, val: T) -> T {
        self.0(val)
    }
}

impl<T: Copy + 'static> From<T> for Call<T> {
    fn from(val: T) -> Call<T> {
        Call(Box::new(move |_| val))
    }
}

impl<F: 'static> From<F> for Call<usize> where F: Fn(usize) -> usize {
    fn from(val: F) -> Call<usize> {
        Call(Box::new(val))
    }
}

impl<F: 'static> From<F> for Call<f64> where F: Fn(f64) -> f64 {
    fn from(val: F) -> Call<f64> {
        Call(Box::new(val))
    }
}
