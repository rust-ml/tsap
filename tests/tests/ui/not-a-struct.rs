use tsap::param;

#[param]
enum Param {
    NotAStruct(usize),
}

fn main() {}
