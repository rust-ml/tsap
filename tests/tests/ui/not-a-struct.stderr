error: item is not a struct

         = help: `#[params]` can only be used on structs

 --> tests/ui/not-a-struct.rs:4:1
  |
4 | / enum Param {
5 | |     NotAStruct(usize),
6 | | }
  | |_^
